### PR TITLE
[mesos/docker] Enhance kube-up to better support running in a container (for CI)

### PR DESCRIPTION
**Pull request by karlkfi**
- Generate CA & API Server SSL key/cert in keygen docker image
  - Refactor SSL generation
  - Generate service account key & user files on local machine
- Enable kube-up to be run in a container (kubernetes-mesos-test)
- Add timeout env vars
  - KUBE_KEYGEN_TIMEOUT (default: 60s)
  - MESOS_DOCKER_ETCD_TIMEOUT (default: 60s)
  - MESOS_DOCKER_MESOS_TIMEOUT (default: 60s)
  - MESOS_DOCKER_API_TIMEOUT (default: 180s)
  - MESOS_DOCKER_ADDON_TIMEOUT (default: 180s)
- Pull docker images up front to avoid timeouts
- Remove docker image builds from test-setup
- Nuke logs dir before each kube-up
- Make run_in_docker work without KUBECONFIG defined
- Fix temp dir cleanup
- Add MESOS_DOCKER_WORK_DIR env var
  - Usage: logs, auth, mesos slave workspaces
  - Default to $HOME/tmp/kubernetes
  - Outside of repo (which gets docker mounted when using kubernetes-mesos-test)
  - Inside $HOME (which gets vm mounted when using docker-machine or boot2docker)
- Add MESOS_DOCKER_WORK_DIR env var
  - Usage: docker-in-docker image storage
  - Default to /var/tmp/kubernetes (survive host reboots)
  - Outside of repo (which gets docker mounted when using kubernetes-mesos-test)
  - Outside $HOME (which gets vm mounted when using docker-machine or boot2docker)
- Enable errtrace
- Increase log level to aid CI debugging
- Improve mesos slave missing error in detect-minions

Fixes `make tests_e2e` run in `mesosphere/kubernetes-mesos-test`.

Mesos-only cluster deployment changes.

xref: https://github.com/mesosphere/kubernetes-mesos/issues/305

### CODE REVIEW

**Comment by karlkfi** [comment]
rebased

**Comment by karlkfi** [comment]
Looks like `test-setup` is called after `kube-up.sh` in `e2e-up.sh`. So creating the images there is pointless, since kube-up already builds them.

Somewhere we need to nuke the `cluster/mesos/docker/logs` dir, but since kube-up populates it on ERR, it should probably be done at the beginning of kube-up.

I've also found that pulling the docker images on a fresh machine can cause the `docker-compose up` to timeout. So I'll probably merge these 3 issues into one PR.

**Comment by karlkfi** [comment]
Added env vars to allow users (e.g. CI) with systems/networks to adjust timeouts.

**Comment by karlkfi** [comment]
Flakey test?

```
Summarizing 1 Failure:

[Fail] Services [It] should serve multiport endpoints from pods 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:1132

Ran 84 of 142 Specs in 384.295 seconds
FAIL! -- 83 Passed | 1 Failed | 2 Pending | 56 Skipped 
```

```
• Failure [118.709 seconds]
Services
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:952
  should serve multiport endpoints from pods [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:235

  Aug 10 20:33:25.638: Timed out waiting for service multi-endpoint-test in namespace e2e-tests-service-0-410m4 to expose endpoints map[podname1:[100] podname2:[101]] (1m0s elapsed)

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:1132
```

**Comment by karlkfi** [comment]
Mesosphere CI still doesn't even finish kube-up with these changes:

```
Waiting (up to 300s) for /var/run/kubernetes/auth/apiserver.crt to exist
Timed out
```

**Comment by jdef** [comment]
lgtm, though i haven't tested this. is there a process diagram somewhere to explain how everything is chained together? it keeps getting more complicated.. :)

File: cluster/mesos/docker/docker-compose.yml [hunk 1]
```diff
@@ -149,9 +156,29 @@ scheduler:
...
+sslgenapiserver:
+  image: mesosphere/kubernetes-mesos-sslgen
+  entrypoint: [ "k8sm-apiserver-sslgen.sh" ]
+  command: [ "/var/run/kubernetes/auth" ]
+  environment:
+  - MESOS_DOCKER_SSLGEN_TIMEOUT
+  volumes:
+  - ${MESOS_DOCKER_VOLUME_ROOT}/auth:/var/run/kubernetes/auth
+  links:
+  - apiserver
```
Comment (tsenart) [inline cluster/mesos/docker/docker-compose.yml]: New line?

File: cluster/mesos/docker/sslgen/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -qqy \
+        curl \
+        openssl \
+        && \
+    apt-get clean
+
+COPY ./bin/* /usr/local/bin/
+
+ENTRYPOINT ["k8sm-sslgen.sh"]
+CMD ["/var/tmp/kubernetes/auth"]
```
Comment (tsenart) [inline cluster/mesos/docker/sslgen/Dockerfile]: New line?

**Comment by tsenart** [comment]
LGTM as far as I can tell. Although I'm quite terrified of these massive amounts of shell scripts.
Would we envision a future where we could use Go for these things? For instance, the crypto packages are quite good and I wonder if they'd be able to match up for these requirements.

**Comment by karlkfi** [comment]
Having slept on it, I think I get to be my own worst enemy.

Pros:
- `docker-compose up` works with minimal setup
- Generating SSL in a container allows a controllable openssl version
- MESOS_DOCKER_VOLUME_ROOT adds flexibility to work around the annoying fact that docker-machine and boot2docker only share `/Users` with their host.
- Configurable timeouts allow short timeouts for dev usage and longer for CI

Cons:
- Adding two containers that run once and die into a docker-compose config feels wrong
  - Hard to read the flow
  - Confusing to a new user that two are required
  - Wrong tool for the job...
- MESOS_DOCKER_VOLUME_ROOT complexity:
  - `/var/run/kubernetes` (in docker-compose.yml config)
  - `/var/tmp/kubernetes` (default in sslgen)
  - `/tmp/kubernetes` (used by CI)
  - `${HOME}/tmp/kubernetes` (used by devs)
- Too many timeouts to configure
- More docker images to build and ship around

File: cluster/mesos/docker/docker-compose.yml [hunk 3]
```diff
@@ -110,20 +114,23 @@ controller:
...
     --cloud-config=/opt/mesos-cloud.conf
     --service-account-private-key-file=/var/run/kubernetes/auth/service-accounts.key
     --root-ca-file=/var/run/kubernetes/auth/root-ca.crt
     --v=2
+  environment:
+  - MESOS_DOCKER_MESOS_TIMEOUT
+  - MESOS_DOCKER_API_TIMEOUT
   volumes:
-  - ./certs/controller:/var/run/kubernetes/auth
+  - ${MESOS_DOCKER_VOLUME_ROOT}/auth:/var/run/kubernetes/auth:ro
```
Comment (sttts) [inline cluster/mesos/docker/docker-compose.yml]: Nice with the new env interpolation in docker-compose.

File: cluster/mesos/docker/docker-compose.yml [hunk 4]
```diff
@@ -149,9 +156,29 @@ scheduler:
...
+  - MESOS_DOCKER_ETCD_TIMEOUT
+  - MESOS_DOCKER_MESOS_TIMEOUT
+  - MESOS_DOCKER_API_TIMEOUT
   links:
   - etcd
   - mesosmaster1
   - mesosslave1
   - mesosslave2
   - apiserver
+sslgen:
```
Comment (sttts) [inline cluster/mesos/docker/docker-compose.yml]: funny to see this. I had the same approach before moving everything into util.sh.

File: cluster/mesos/docker/docker-compose.yml [hunk 5]
```diff
@@ -149,9 +156,29 @@ scheduler:
...
+  image: mesosphere/kubernetes-mesos-sslgen
+  entrypoint: [ "k8sm-sslgen.sh" ]
+  command: [ "/var/run/kubernetes/auth" ]
+  volumes:
+  - ${MESOS_DOCKER_VOLUME_ROOT}/auth:/var/run/kubernetes/auth
```
Comment (sttts) [inline cluster/mesos/docker/docker-compose.yml]: Why no volume mount from this container directly?

Comment (sttts) [inline cluster/mesos/docker/docker-compose.yml]: Then we can get rid of the temporary directory altogether

File: cluster/mesos/docker/util.sh [hunk 6]
```diff
@@ -152,68 +160,54 @@ function verify-prereqs {
...
+  # Nuke auth up front so that they get created in the right order
+  #TODO(karlkfi): reuse existing credentials/certs/keys
+  mkdir -p "${auth_dir}"
+  rm -rf "${auth_dir}"/*
+
   if [ "${MESOS_DOCKER_SKIP_BUILD:-false}" != "true" ]; then
+    # Pull before `docker-compose up` to avoid timeouts between container runs.
+    # Pull before building images (but only if built) to avoid overwriting locally built images.
+    echo "Pulling docker images" 1>&2
+    docker-compose -f "${compose_file}" pull
```
Comment (sttts) [inline cluster/mesos/docker/util.sh]: Didn't know this is possible. :+1:

File: cluster/mesos/docker/docker-compose.yml (hunk 5 above)
Comment (karlkfi) [inline cluster/mesos/docker/docker-compose.yml]: kubectl configuration still needs access to the `root-ca.crt` and `token-users`, but now that I think about it, we can probably use `docker cp` to get them out.

File: cluster/mesos/docker/util.sh (hunk 6 above)
Comment (karlkfi) [inline cluster/mesos/docker/util.sh]: Downside is that it pulls whether you have a newer one built locally or not.

**Comment by karlkfi** [comment]
rebased to fix shippable error

**Comment by karlkfi** [comment]
Squashed.

**Comment by karlkfi** [comment]
@zmerlynn You wanna give this bashiness a look-see?

**Comment by zmerlynn** [comment]
Sorry, I didn't see this. Will look soon.

**Comment by karlkfi** [comment]
@zmerlynn You gonna have a chance to take a look at this or should I just get @sttts to give it the ok?
It won't affect anything outside of the mesos/docker cluster. I mostly just roped you in to increase the bus factor and get some feedback on consistency with other cluster's bash scripts. The containerized SSL cert generation, for example, might be useful elsewhere.

**Comment by zmerlynn** [comment]
Ack, sorry! I'll look today.

**Comment by karlkfi** [comment]
Thanks, @zmerlynn !

**Pull request merged**